File: algorithms/cpp/074_search_a_2D_matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        return binarySearch(matrix,0, matrix.size()*matrix[0].size()-1,target);
        
    }
    bool binarySearch(vector<vector<int> >& nums, int left, int right, int target) {
        if (left==right) 
            return (nums[left/nums[0].size()][left%nums[0].size()]==target);
        else if (left>right) return false;
        int mid=(left+right)/2;
        int row_idx=mid/nums[0].size(), col_idx=mid%nums[0].size();
        if (nums[row_idx][col_idx]==target) return true;
        else if (nums[row_idx][col_idx]>target) return binarySearch (nums,left,mid-1,target);
        else return binarySearch(nums,mid+1,right,target);
    }

};

File: algorithms/python3/074_search_a_2d_matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        r=len(matrix)
        if not r:
            return False
        c=len(matrix[0])
        start,end=0,r*c-1
        while start<=end:
            mid=(start+end)//2
            curr,curc=mid//c,mid%c
            if matrix[curr][curc]==target:
                return True
            elif matrix[curr][curc]>target:
                end=mid-1
            else:
                start=mid+1
        return False


a=Solution()

nums=matrix = [
  [1,   3,  5,  7],
  [10, 11, 16, 20],
  [23, 30, 34, 50]
]

print(a.searchMatrix(nums,3))